### ISSUE TITLE
character count issues with foreign akphabets

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1100]
Hi, if you type or paste workds from different alphabets in Pd, like katakana, you don't get a proper box size.

For instance, let's take the word for karaoke (カラオケ), you have 4 "letters". Pd is actually good in counting it, so it resizes as if you had 4 characters in the latin language, which doesn't really work as it actually takes more space. So you end with a shorter box size for any box type (message, comment, symbol box, object). See image below.

<img width="346" alt="Screen Shot 2020-07-28 at 01 14 29" src="https://user-images.githubusercontent.com/5360142/88618516-b67fd680-d06f-11ea-8506-bac8d366bd1e.png">

Note however that I also have cyclone comment in the example and it does get the size correctly

**Comment by porres** [comment]
by the way, I'm running into a similar issue with a vanilla abstracton where I use a cnv's label to display symbols, and I count the number of characters but it doesn't match katakana... so I wouldn't know how to fix this in this vanilla abstraction, unless there is a way to know the size of characters from katakana and other alphabets.

<img width="349" alt="Screen Shot 2020-07-28 at 01 41 29" src="https://user-images.githubusercontent.com/5360142/88620247-e4ffb080-d073-11ea-869c-26e864c8e550.png">

**Comment by Spacechild1** [comment]
In you compare latin and katakana side by side you'll notice that each katakana character takes the space of two latin characters. Japanese kana/kanji need a square space, so the typical space for latin characters is too narrow. I think there's no way to fix this... Pd simply can't know which unicode characters take more space.

> Note however that I also have cyclone comment in the example and it does get the size correctly

How do you count characters in `[cyclone/comment]`? ;-)

**Comment by porres** [comment]
> How do you count characters in [cyclone/comment]? ;-)

In the same way as pd vanilla's comment does. But as for getting the size of the object, then it's a different thing... there's a tcl callback function that just gives us the size of the object.

**Comment by Spacechild1** [comment]
Ah, I see, `[cyclone/comment]` uses `bbox` https://www.tcl.tk/man/tcl8.4/TkCmd/canvas.htm#M37 to get the actual bounding box as determined by Tcl. Pd simply takes the number of unicode code points (using the UTF8 library in `s_utf8.h`). This definitely explains the different results.

**Comment by umlaeute** [comment]
so probably Pd should use `bbox` as well.
(as in: let the GUI report back how big the created object actually is, rather than assume to know better)

**Comment by porres** [comment]
I guess so :) and if you select the text anyway, you access the actual size, cause deep inside, it knows.

<img width="138" alt="Screen Shot 2020-07-28 at 13 42 08" src="https://user-images.githubusercontent.com/5360142/88695352-275ad900-d0d8-11ea-8d26-f57675b06e07.png">

**Comment by umlaeute** [comment]
> yeah, if you select the text anyway, you access the actual size, cause deep inside, it knows.

nope. the selection is handled by TclTk (which of course knows the size), not by Pd.

**Comment by porres** [comment]
that's what I meant, the GUI knows ;)

**Comment by umlaeute** [comment]
well yea. that's why `bbox` works...

anyhow, could you fix the title? this is not about "counting characters" but about "wrongly calculating the text-width" ("counting characters" is a naive - and obviously borked - attempt to achieve the latter)

**Comment by umlaeute** [comment]
speaking of `bbox`, @porres do you get pixel-identical sizes on all platforms with `[cylcone/comment]`?

**Comment by Spacechild1** [comment]
The danger of using `bbox` is to rely even more on Tcl/Tk.

>  @porres do you get pixel-identical sizes on all platforms with [cylcone/comment]?

Good question. Also: is the result of `bbox` guaranteed to be stable across Tcl/Tk versions?

Maybe we could use the result of `bbox` only as a hint.

**Comment by porres** [comment]
I have no reason to suspect it isn't consistent cross platform. Checking the help file, it all looks the same, it looks more consistent that what I already see with opening the same patch in different platforms, where vanilla comments don't seem to render at exactly the same size.

Anyway, I wanted to print bbox values and test but couldn't compile my new version of cyclone/comment on linux/windows, see https://github.com/pure-data/pure-data/issues/1108

**Comment by porres** [comment]
ok, so I got slightly different values, and they might be already the case with rendering text in different platforms as I also find a similar issue when opening patches in different platforms. For example, if on the mac there's a one pixel horizontal difference between two comments (one on the top of the other), I won't see the same separation on Linux because of this slight difference.

what I'm trying to say is that there are already some issues in getting exact font sizes in Pd as it is. Here's the screenshot with the different values returned by bbox in each platform 

<img width="447" alt="Screen Shot 2020-07-29 at 23 02 34" src="https://user-images.githubusercontent.com/5360142/88872218-14d3c300-d1f0-11ea-9ce0-62f57b7cc45b.png">

**Comment by porres** [comment]
dimensions in each platform

windows: width 352, heigth 35
mac: width 354, height 34
linux: width 357, height 35

**Comment by porres** [comment]
I have to redo the test,  i realize I wasn't using "dejavu sans mono"

**Comment by porres** [comment]
Ok, using DejaVu Sans Mono now (DVSM), font size 30, results in the screenshot. Yeah, I'm also getting slightly different values.

<img width="644" alt="Screen Shot 2020-07-30 at 00 27 03" src="https://user-images.githubusercontent.com/5360142/88877278-7e0d0380-d1fb-11ea-8723-198d602305a6.png">

**Comment by porres** [comment]
you can see I'm also comparing with a [cnv] object that has a lable with the same text, font and fontsize. I did it just to make sure fonts looked the same and had the same size/width. And they do, so it's not like only cyclone/comment is changing in size, so does [cnv]'s label font in the same way. And, like I said, I also see this issue when it comes to rendering vanilla's comments - things don't render and look exactly the same cross platform.
